File: backend/app/api/v1/endpoints/acceptance_report.py
"""成果验收正式报告端点。"""

from typing import Annotated

from fastapi import APIRouter, Query
from fastapi.responses import FileResponse

from app.api.deps import DatabaseSession
from app.schemas.acceptance_report import (
    AcceptanceReportGenerateRequest,
    AcceptanceReportListResponse,
    AcceptanceReportResponse,
)
from app.services.acceptance_report_service import AcceptanceReportService

router = APIRouter(
    prefix="/api/v1/acceptance-reports",
    tags=["成果验收报告"],
)
service = AcceptanceReportService()


@router.get("", response_model=AcceptanceReportListResponse)
async def list_acceptance_reports(
    db: DatabaseSession,
    task_code: Annotated[str, Query(min_length=1, max_length=50)] = "RS-2026-045",
) -> AcceptanceReportListResponse:
    """查询任务验收报告门禁与版本历史。

    Args:
        db: FastAPI 注入的异步数据库会话。
        task_code: 作业任务编号。

    Returns:
        AcceptanceReportListResponse: 生成条件和报告版本。
    """
    return await service.list_reports(db, task_code)


@router.post("/generate", response_model=AcceptanceReportResponse)
async def generate_acceptance_report(
    request: AcceptanceReportGenerateRequest,
    db: DatabaseSession,
    task_code: Annotated[str, Query(min_length=1, max_length=50)] = "RS-2026-045",
) -> AcceptanceReportResponse:
    """基于当前有效成果包生成 DOCX/PDF 验收报告。

    Args:
        request: 报告标题、操作人和生成依据。
        db: FastAPI 注入的异步数据库会话。
        task_code: 作业任务编号。

    Returns:
        AcceptanceReportResponse: 新生成报告摘要。
    """
    return await service.generate_report(db, task_code, request)


@router.get("/{report_code}/download", response_class=FileResponse)
async def download_acceptance_report(
    report_code: str,
    db: DatabaseSession,
    requester_code: Annotated[str, Query(min_length=1, max_length=50)],
) -> FileResponse:
    """鉴权并完整复核后下载验收报告 ZIP。

    Args:
        report_code: 验收报告编号。
        db: FastAPI 注入的异步数据库会话。
        requester_code: 下载人稳定项目用户编码。

    Returns:
        FileResponse: 包含 DOCX、PDF 和 manifest 的 ZIP。
    """
    download = await service.authorize_download(
        db,
        report_code,
        requester_code,
    )
    return FileResponse(
        path=download.path,
        filename=download.filename,
        media_type="application/zip",
        headers={"ETag": f'"{download.checksum_sha256}"'},
    )

File: frontend/src/api/acceptanceReport.ts
import request from '@/api/request'
import type {
  AcceptanceReport,
  AcceptanceReportGeneratePayload,
  AcceptanceReportList,
} from '@/types/acceptanceReport'

/**
 * 查询任务成果验收正式报告门禁和版本历史。
 * @param taskCode 作业任务编号。
 * @returns 当前业务门禁和报告版本。
 */
export const getAcceptanceReports = (
  taskCode: string,
): Promise<AcceptanceReportList> => request.get('/v1/acceptance-reports', {
  params: { task_code: taskCode },
  timeout: 60_000,
})

/**
 * 基于当前有效成果包生成 DOCX/PDF 验收报告。
 * @param payload 报告标题、生成依据和稳定用户编码。
 * @param taskCode 作业任务编号。
 * @returns 新生成验收报告。
 */
export const generateAcceptanceReport = (
  payload: AcceptanceReportGeneratePayload,
  taskCode: string,
): Promise<AcceptanceReport> => request.post(
  '/v1/acceptance-reports/generate',
  payload,
  { params: { task_code: taskCode }, timeout: 300_000 },
)

/**
 * 下载服务端重新校验的 DOCX/PDF 验收报告 ZIP。
 * @param reportCode 验收报告编号。
 * @param requesterCode 下载人稳定编码。
 * @returns 报告 ZIP 实体。
 */
export const downloadAcceptanceReport = (
  reportCode: string,
  requesterCode: string,
): Promise<Blob> => request.get(
  `/v1/acceptance-reports/${encodeURIComponent(reportCode)}/download`,
  {
    params: { requester_code: requesterCode },
    responseType: 'blob',
    timeout: 300_000,
  },
)
